ccl/changefeedccl: TestNoBackfillAfterNonTargetColumnDrop failed

ccl/changefeedccl.TestNoBackfillAfterNonTargetColumnDrop [failed](https://tanzanite.cluster.engflow.com/invocations/default/17d21dc5-0300-465e-917e-bc6841dcc85f?testReportRun=3&testReportShard=6&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [c712dbafb5ab5ed4203f342274a76bcf13eec686](https://github.com/cockroachdb/cockroach/commits/c712dbafb5ab5ed4203f342274a76bcf13eec686):


```
test logs left over in: outputs.zip/logTestNoBackfillAfterNonTargetColumnDrop104444361
--- FAIL: TestNoBackfillAfterNonTargetColumnDrop (16.33s)
=== RUN   TestNoBackfillAfterNonTargetColumnDrop/sinkless
    helpers_test.go:1038: making server as secondary tenant
    helpers_test.go:1117: making sinkless feed factory
    helpers_test.go:1182: pgURL sinkless SinklessFeedUser
    helpers_test.go:1182: pgURL sinkless root
    testfeed_test.go:291: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE hasfams FAMILY b_and_c
    changefeed_test.go:1922: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:266
        	            				pkg/ccl/changefeedccl/helpers_test.go:361
        	            				pkg/ccl/changefeedccl/changefeed_test.go:1922
        	            				pkg/ccl/changefeedccl/helpers_test.go:1289
        	            				pkg/ccl/changefeedccl/helpers_test.go:1324
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:124
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:283
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:269
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:354
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:350
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:267
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:361
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestNoBackfillAfterNonTargetColumnDrop.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:1922
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1289
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1324
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1690
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestNoBackfillAfterNonTargetColumnDrop/sinkless
    testfeed_test.go:298: closing sinkless feed
--- FAIL: TestNoBackfillAfterNonTargetColumnDrop/sinkless (16.33s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestNoBackfillAfterNonTargetColumnDrop.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49197

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -867,6 +867,10 @@ func (ca *changeAggregator) flushBufferedEvents() error {
 // changeAggregator node to the changeFrontier node to allow the changeFrontier
 // to persist the overall changefeed's progress
 func (ca *changeAggregator) noteResolvedSpan(resolved jobspb.ResolvedSpan) (returnErr error) {
+	if log.V(2) {
+		log.Infof(ca.Ctx(), "resolved span from kv feed: %#v", resolved)
+	}
+
 	if resolved.Timestamp.IsEmpty() {
 		// @0.0 resolved timestamps could come in from rangefeed checkpoint.
 		// When rangefeed starts running, it emits @0.0 resolved timestamp.
@@ -949,6 +953,9 @@ func (ca *changeAggregator) emitResolved(batch jobspb.ResolvedSpans) error {
 		Stats: jobspb.ResolvedSpans_Stats{
 			RecentKvCount: ca.recentKVCount,
 		},
+	}
+	if log.V(2) {
+		log.Infof(ca.Ctx(), "progress update to be sent to change frontier: %#v", progressUpdate)
 	}
 	updateBytes, err := protoutil.Marshal(&progressUpdate)
 	if err != nil {

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -878,7 +878,7 @@ func (ca *changeAggregator) noteResolvedSpan(resolved jobspb.ResolvedSpan) (retu
 		return nil
 	}
 
-	advanced, err := ca.frontier.ForwardResolvedSpan(ca.Ctx(), resolved)
+	advanced, err := ca.frontier.ForwardResolvedSpan(resolved)
 	if err != nil {
 		return err
 	}
@@ -1660,7 +1660,7 @@ func (cf *changeFrontier) noteAggregatorProgress(d rowenc.EncDatum) error {
 }
 
 func (cf *changeFrontier) forwardFrontier(resolved jobspb.ResolvedSpan) error {
-	frontierChanged, err := cf.frontier.ForwardResolvedSpan(cf.Ctx(), resolved)
+	frontierChanged, err := cf.frontier.ForwardResolvedSpan(resolved)
 	if err != nil {
 		return err
 	}

--- a/pkg/ccl/changefeedccl/resolvedspan/BUILD.bazel
+++ b/pkg/ccl/changefeedccl/resolvedspan/BUILD.bazel
@@ -15,7 +15,6 @@ go_library(
         "//pkg/roachpb",
         "//pkg/settings",
         "//pkg/util/hlc",
-        "//pkg/util/log",
         "//pkg/util/span",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_redact//:redact",

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier.go
@@ -15,7 +15,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
-	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/span"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
@@ -51,7 +50,7 @@ func (f *AggregatorFrontier) ForwardResolvedSpan(
 		// Boundary resolved events should be ingested from the schema feed
 		// serially, where the changefeed won't ever observe a new schema change
 		// boundary until it has progressed past the current boundary.
-		if err := f.assertBoundaryNotEarlier(ctx, r); err != nil {
+		if err := f.assertBoundaryNotEarlierOrDifferent(r); err != nil {
 			return false, err
 		}
 	default:
@@ -102,13 +101,14 @@ func (f *CoordinatorFrontier) ForwardResolvedSpan(
 		// it is a BACKFILL we have already seen, then it is fine for it to be
 		// an earlier timestamp than the latest boundary.
 		boundaryTS := r.Timestamp
-		_, ok := slices.BinarySearchFunc(f.backfills, boundaryTS, func(elem hlc.Timestamp, ts hlc.Timestamp) int {
-			return elem.Compare(ts)
-		})
-		if ok {
+		if _, ok := slices.BinarySearchFunc(f.backfills, boundaryTS,
+			func(elem hlc.Timestamp, ts hlc.Timestamp) int {
+				return elem.Compare(ts)
+			},
+		); ok {
 			break
 		}
-		if err := f.assertBoundaryNotEarlier(ctx, r); err != nil {
+		if err := f.assertBoundaryNotEarlierOrDifferent(r); err != nil {
 			return false, err
 		}
 		f.backfills = append(f.backfills, boundaryTS)
@@ -116,7 +116,7 @@ func (f *CoordinatorFrontier) ForwardResolvedSpan(
 		// EXIT and RESTART are final boundaries that cause the changefeed
 		// processors to all move to draining and so should not be followed
 		// by any other boundaries.
-		if err := f.assertBoundaryNotEarlier(ctx, r); err != nil {
+		if err := f.assertBoundaryNotEarlierOrDifferent(r); err != nil {
 			return false, err
 		}
 	default:
@@ -129,9 +129,10 @@ func (f *CoordinatorFrontier) ForwardResolvedSpan(
 	// If the frontier changed, we check if the frontier has advanced past any known backfills.
 	if frontierChanged {
 		frontier := f.Frontier()
-		i, _ := slices.BinarySearchFunc(f.backfills, frontier, func(elem hlc.Timestamp, ts hlc.Timestamp) int {
-			return elem.Compare(ts)
-		})
+		i, _ := slices.BinarySearchFunc(f.backfills, frontier,
+			func(elem hlc.Timestamp, ts hlc.Timestamp) int {
+				return elem.Compare(ts)
+			})
 		f.backfills = f.backfills[i:]
 	}
 	return frontierChanged, nil
@@ -144,10 +145,11 @@ func (f *CoordinatorFrontier) ForwardResolvedSpan(
 // happening at different timestamps.
 func (f *CoordinatorFrontier) InBackfill(r jobspb.ResolvedSpan) bool {
 	boundaryTS := r.Timestamp
-	_, ok := slices.BinarySearchFunc(f.backfills, boundaryTS, func(elem hlc.Timestamp, ts hlc.Timestamp) int {
-		return elem.Next().Compare(ts)
-	})
-	if ok {
+	if _, ok := slices.BinarySearchFunc(f.backfills, boundaryTS,
+		func(elem hlc.Timestamp, ts hlc.Timestamp) int {
+			return elem.Next().Compare(ts)
+		},
+	); ok {
 		return true
 	}
 
@@ -160,11 +162,12 @@ func (f *CoordinatorFrontier) All() iter.Seq[jobspb.ResolvedSpan] {
 		for resolvedSpan := range f.resolvedSpanFrontier.All() {
 			// Check if it's at an earlier backfill boundary.
 			if resolvedSpan.BoundaryType == jobspb.ResolvedSpan_NONE {
-				for _, b := range f.backfills {
-					if b.Equal(resolvedSpan.Timestamp) {
-						resolvedSpan.BoundaryType = jobspb.ResolvedSpan_BACKFILL
-						break
-					}
+				if _, ok := slices.BinarySearchFunc(f.backfills, resolvedSpan.Timestamp,
+					func(elem hlc.Timestamp, ts hlc.Timestamp) int {
+						return elem.Compare(ts)
+					},
+				); ok {
+					resolvedSpan.BoundaryType = jobspb.ResolvedSpan_BACKFILL
 				}
 			}
 			if !yield(resolvedSpan) {
@@ -295,23 +298,26 @@ func (f *resolvedSpanFrontier) InBackfill(r jobspb.ResolvedSpan) bool {
 	return false
 }
 
-// assertBoundaryNotEarlier is a helper method provided to assert that a
-// resolved span does not have an earlier boundary than the existing one.
-func (f *resolvedSpanFrontier) assertBoundaryNotEarlier(
-	ctx context.Context, r jobspb.ResolvedSpan,
-) error {
+// assertBoundaryNotEarlierOrDifferent is a helper method that asserts that a
+// resolved span does not have an earlier boundary than the existing one
+// nor is it at the same time as the existing one with a different type.
+func (f *resolvedSpanFrontier) assertBoundaryNotEarlierOrDifferent(r jobspb.ResolvedSpan) error {
 	boundaryType := r.BoundaryType
 	if boundaryType == jobspb.ResolvedSpan_NONE {
-		return errors.AssertionFailedf("assertBoundaryNotEarlier should not be called for NONE boundary")
+		return errors.AssertionFailedf(
+			"assertBoundaryNotEarlierOrDifferent should not be called for NONE boundary")
 	}
 	boundaryTS := r.Timestamp
+	newBoundary := newResolvedSpanBoundary(boundaryTS, boundaryType)
 	if f.boundary.After(boundaryTS) {
-		newBoundary := newResolvedSpanBoundary(boundaryTS, boundaryType)
-		err := errors.AssertionFailedf("received resolved span for %s "+
+		return errors.AssertionFailedf("received resolved span for %s "+
 			"with %v, which is earlier than previously received %v",
 			r.Span, newBoundary, f.boundary)
-		log.Errorf(ctx, "error while forwarding boundary resolved span: %v", err)
-		return err
+	}
+	if atBoundary, typ := f.boundary.At(boundaryTS); atBoundary && boundaryType != typ {
+		return errors.AssertionFailedf("received resolved span for %s "+
+			"with %v, which has a different type from previously received %v with same timestamp",
+			r.Span, newBoundary, f.boundary)
 	}
 	return nil
 }

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier.go
@@ -6,7 +6,6 @@
 package resolvedspan
 
 import (
-	"context"
 	"iter"
 	"slices"
 
@@ -44,8 +43,8 @@ func NewAggregatorFrontier(
 // ForwardResolvedSpan forwards the progress of a resolved span and also does
 // some boundary validation.
 func (f *AggregatorFrontier) ForwardResolvedSpan(
-	ctx context.Context, r jobspb.ResolvedSpan,
-) (bool, error) {
+	r jobspb.ResolvedSpan,
+) (forwarded bool, err error) {
 	switch boundaryType := r.BoundaryType; boundaryType {
 	case jobspb.ResolvedSpan_NONE:
 	case jobspb.ResolvedSpan_BACKFILL, jobspb.ResolvedSpan_EXIT, jobspb.ResolvedSpan_RESTART:
@@ -89,8 +88,8 @@ func NewCoordinatorFrontier(
 // ForwardResolvedSpan forwards the progress of a resolved span and also does
 // some boundary validation.
 func (f *CoordinatorFrontier) ForwardResolvedSpan(
-	ctx context.Context, r jobspb.ResolvedSpan,
-) (bool, error) {
+	r jobspb.ResolvedSpan,
+) (forwarded bool, err error) {
 	switch boundaryType := r.BoundaryType; boundaryType {
 	case jobspb.ResolvedSpan_NONE:
 	case jobspb.ResolvedSpan_BACKFILL:
@@ -224,8 +223,14 @@ func newResolvedSpanFrontier(
 }
 
 // ForwardResolvedSpan forwards the progress of a resolved span.
-func (f *resolvedSpanFrontier) ForwardResolvedSpan(r jobspb.ResolvedSpan) (bool, error) {
-	forwarded, err := f.Forward(r.Span, r.Timestamp)
+// The frontier is considered forwarded if either the frontier
+// timestamp advances or the current frontier timestamp becomes
+// a boundary timestamp (for some non-NONE boundary type)
+// and all the spans are at the boundary timestamp already.
+func (f *resolvedSpanFrontier) ForwardResolvedSpan(
+	r jobspb.ResolvedSpan,
+) (forwarded bool, err error) {
+	forwarded, err = f.Forward(r.Span, r.Timestamp)
 	if err != nil {
 		return false, err
 	}
@@ -235,7 +240,13 @@ func (f *resolvedSpanFrontier) ForwardResolvedSpan(r jobspb.ResolvedSpan) (bool,
 			ts:  r.Timestamp,
 			typ: r.BoundaryType,
 		}
-		f.boundary.Forward(newBoundary)
+		boundaryForwarded := f.boundary.Forward(newBoundary)
+		if boundaryForwarded && !forwarded {
+			// The frontier is considered forwarded if the boundary type
+			// changes to non-NONE and all the spans are at the boundary
+			// timestamp already.
+			forwarded, _, _ = f.AtBoundary()
+		}
 	}
 	return forwarded, nil
 }
@@ -375,9 +386,9 @@ func (b *resolvedSpanBoundary) At(ts hlc.Timestamp) (bool, jobspb.ResolvedSpan_B
 	return false, 0
 }
 
-// After returns whether a timestamp is later than the boundary timestamp.
+// After returns whether the boundary is after a given timestamp.
 func (b *resolvedSpanBoundary) After(ts hlc.Timestamp) bool {
-	return ts.Less(b.ts)
+	return b.ts.After(ts)
 }
 
 // Forward forwards the boundary to the new boundary if it is later.

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
@@ -155,7 +155,7 @@ func TestCoordinatorFrontier(t *testing.T) {
 
 type frontier interface {
 	Frontier() hlc.Timestamp
-	ForwardResolvedSpan(context.Context, jobspb.ResolvedSpan) (bool, error)
+	ForwardResolvedSpan(jobspb.ResolvedSpan) (bool, error)
 	InBackfill(jobspb.ResolvedSpan) bool
 	AtBoundary() (bool, jobspb.ResolvedSpan_BoundaryType, hlc.Timestamp)
 	All() iter.Seq[jobspb.ResolvedSpan]
@@ -171,7 +171,7 @@ func testBackfillSpan(
 ) {
 	backfillSpan := makeResolvedSpan(start, end, ts, jobspb.ResolvedSpan_NONE)
 	require.True(t, f.InBackfill(backfillSpan))
-	_, err := f.ForwardResolvedSpan(ctx, backfillSpan)
+	_, err := f.ForwardResolvedSpan(backfillSpan)
 	require.NoError(t, err)
 	require.Equal(t, frontierAfterSpan, f.Frontier())
 }
@@ -186,7 +186,7 @@ func testBoundarySpan(
 	frontierAfterSpan hlc.Timestamp,
 ) {
 	boundarySpan := makeResolvedSpan(start, end, boundaryTS, boundaryType)
-	_, err := f.ForwardResolvedSpan(ctx, boundarySpan)
+	_, err := f.ForwardResolvedSpan(boundarySpan)
 	require.NoError(t, err)
 
 	if finalBoundarySpan := frontierAfterSpan.Equal(boundaryTS); finalBoundarySpan {
@@ -220,7 +220,7 @@ func testIllegalBoundarySpan(
 	boundaryType jobspb.ResolvedSpan_BoundaryType,
 ) {
 	boundarySpan := makeResolvedSpan(start, end, boundaryTS, boundaryType)
-	_, err := f.ForwardResolvedSpan(ctx, boundarySpan)
+	_, err := f.ForwardResolvedSpan(boundarySpan)
 	require.True(t, errors.HasAssertionFailure(err))
 }
 
@@ -243,4 +243,54 @@ func makeSpan(start, end string) roachpb.Span {
 		Key:    roachpb.Key(start),
 		EndKey: roachpb.Key(end),
 	}
+}
+
+func TestAggregatorFrontier_ForwardResolvedSpan(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// Create a fresh frontier with no progress.
+	f, err := resolvedspan.NewAggregatorFrontier(
+		hlc.Timestamp{},
+		hlc.Timestamp{},
+		makeSpan("a", "f"),
+	)
+	require.NoError(t, err)
+	require.Zero(t, f.Frontier())
+
+	t.Run("advance frontier with no boundary", func(t *testing.T) {
+		// Forwarding part of the span space to 10 should not advance the frontier.
+		forwarded, err := f.ForwardResolvedSpan(
+			makeResolvedSpan("a", "b", makeTS(10), jobspb.ResolvedSpan_NONE))
+		require.NoError(t, err)
+		require.False(t, forwarded)
+		require.Zero(t, f.Frontier())
+
+		// Forwarding the rest of the span space to 10 should advance the frontier.
+		forwarded, err = f.ForwardResolvedSpan(
+			makeResolvedSpan("b", "f", makeTS(10), jobspb.ResolvedSpan_NONE))
+		require.NoError(t, err)
+		require.True(t, forwarded)
+		require.Equal(t, makeTS(10), f.Frontier())
+	})
+
+	t.Run("advance frontier with same timestamp and new boundary", func(t *testing.T) {
+		// Forwarding part of the span space to 10 again with a non-NONE boundary
+		// should be considered forwarding the frontier because we're learning
+		// about a new boundary.
+		forwarded, err := f.ForwardResolvedSpan(
+			makeResolvedSpan("c", "f", makeTS(10), jobspb.ResolvedSpan_RESTART))
+		require.NoError(t, err)
+		require.True(t, forwarded)
+		require.Equal(t, makeTS(10), f.Frontier())
+
+		// Forwarding the rest of the span space to 10 again with a non-NONE boundary
+		// should not be considered forwarding the frontier because we already
+		// know about the new boundary.
+		forwarded, err = f.ForwardResolvedSpan(
+			makeResolvedSpan("a", "c", makeTS(10), jobspb.ResolvedSpan_RESTART))
+		require.NoError(t, err)
+		require.False(t, forwarded)
+		require.Equal(t, makeTS(10), f.Frontier())
+	})
 }

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
@@ -6,7 +6,6 @@
 package resolvedspan_test
 
 import (
-	"context"
 	"iter"
 	"testing"
 
@@ -24,8 +23,6 @@ func TestAggregatorFrontier(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	ctx := context.Background()
-
 	// Create a fresh frontier with no progress.
 	statementTime := makeTS(10)
 	var initialHighwater hlc.Timestamp
@@ -38,32 +35,37 @@ func TestAggregatorFrontier(t *testing.T) {
 	require.Equal(t, initialHighwater, f.Frontier())
 
 	// Forward spans representing initial scan.
-	testBackfillSpan(t, ctx, f, "a", "b", statementTime, initialHighwater)
-	testBackfillSpan(t, ctx, f, "b", "f", statementTime, statementTime)
+	testBackfillSpan(t, f, "a", "b", statementTime, initialHighwater)
+	testBackfillSpan(t, f, "b", "f", statementTime, statementTime)
 
 	// Forward spans signalling a backfill is required.
 	backfillTS := makeTS(20)
-	testBoundarySpan(t, ctx, f, "a", "b", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
-	testBoundarySpan(t, ctx, f, "b", "c", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
-	testBoundarySpan(t, ctx, f, "c", "d", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
-	testBoundarySpan(t, ctx, f, "d", "e", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
-	testBoundarySpan(t, ctx, f, "e", "f", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, backfillTS.Prev())
+	testBoundarySpan(t, f, "a", "b", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "b", "c", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "c", "d", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "d", "e", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "e", "f", backfillTS.Prev(), jobspb.ResolvedSpan_BACKFILL, backfillTS.Prev())
 
-	// Confirm that attempting to signal an earlier boundary causes an assertion error.
+	// Verify that attempting to signal an earlier boundary causes an assertion error.
 	illegalBoundaryTS := makeTS(15)
-	testIllegalBoundarySpan(t, ctx, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_BACKFILL)
-	testIllegalBoundarySpan(t, ctx, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_RESTART)
-	testIllegalBoundarySpan(t, ctx, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_EXIT)
+	testIllegalBoundarySpan(t, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_BACKFILL)
+	testIllegalBoundarySpan(t, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_RESTART)
+	testIllegalBoundarySpan(t, f, "a", "f", illegalBoundaryTS, jobspb.ResolvedSpan_EXIT)
+
+	// Verify that attempting to signal a boundary at the latest boundary time with a different
+	// boundary type causes an assertion error.
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS.Prev(), jobspb.ResolvedSpan_RESTART)
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS.Prev(), jobspb.ResolvedSpan_EXIT)
 
 	// Forward spans representing actual backfill.
-	testBackfillSpan(t, ctx, f, "d", "e", backfillTS, backfillTS.Prev())
-	testBackfillSpan(t, ctx, f, "e", "f", backfillTS, backfillTS.Prev())
-	testBackfillSpan(t, ctx, f, "a", "d", backfillTS, backfillTS)
+	testBackfillSpan(t, f, "d", "e", backfillTS, backfillTS.Prev())
+	testBackfillSpan(t, f, "e", "f", backfillTS, backfillTS.Prev())
+	testBackfillSpan(t, f, "a", "d", backfillTS, backfillTS)
 
 	// Forward spans signalling a restart is required.
 	restartTS := makeTS(30)
-	testBoundarySpan(t, ctx, f, "a", "b", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, backfillTS)
-	testBoundarySpan(t, ctx, f, "b", "f", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, restartTS.Prev())
+	testBoundarySpan(t, f, "a", "b", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, backfillTS)
+	testBoundarySpan(t, f, "b", "f", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, restartTS.Prev())
 
 	// Simulate restarting by creating a new frontier with the initial highwater
 	// set to the previous frontier timestamp.
@@ -76,20 +78,18 @@ func TestAggregatorFrontier(t *testing.T) {
 	require.NoError(t, err)
 
 	// Forward spans representing post-restart backfill.
-	testBackfillSpan(t, ctx, f, "a", "b", restartTS, initialHighwater)
-	testBackfillSpan(t, ctx, f, "e", "f", restartTS, initialHighwater)
-	testBackfillSpan(t, ctx, f, "b", "e", restartTS, restartTS)
+	testBackfillSpan(t, f, "a", "b", restartTS, initialHighwater)
+	testBackfillSpan(t, f, "e", "f", restartTS, initialHighwater)
+	testBackfillSpan(t, f, "b", "e", restartTS, restartTS)
 
 	// Forward spans signalling an exit is required.
 	exitTS := makeTS(40)
-	testBoundarySpan(t, ctx, f, "a", "f", exitTS.Prev(), jobspb.ResolvedSpan_EXIT, exitTS.Prev())
+	testBoundarySpan(t, f, "a", "f", exitTS.Prev(), jobspb.ResolvedSpan_EXIT, exitTS.Prev())
 }
 
 func TestCoordinatorFrontier(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-
-	ctx := context.Background()
 
 	// Create a fresh frontier with no progress.
 	statementTime := makeTS(10)
@@ -103,35 +103,40 @@ func TestCoordinatorFrontier(t *testing.T) {
 	require.Equal(t, initialHighwater, f.Frontier())
 
 	// Forward spans representing initial scan.
-	testBackfillSpan(t, ctx, f, "a", "b", statementTime, initialHighwater)
-	testBackfillSpan(t, ctx, f, "b", "f", statementTime, statementTime)
+	testBackfillSpan(t, f, "a", "b", statementTime, initialHighwater)
+	testBackfillSpan(t, f, "b", "f", statementTime, statementTime)
 
 	// Forward span signalling a backfill is required.
 	backfillTS1 := makeTS(15)
-	testBoundarySpan(t, ctx, f, "a", "b", backfillTS1.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "a", "b", backfillTS1.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
 
 	// Forward span signalling another backfill is required (simulates multiple
 	// aggregators progressing at different speeds).
 	backfillTS2 := makeTS(20)
-	testBackfillSpan(t, ctx, f, "a", "b", backfillTS1, statementTime)
-	testBoundarySpan(t, ctx, f, "a", "b", backfillTS2.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBackfillSpan(t, f, "a", "b", backfillTS1, statementTime)
+	testBoundarySpan(t, f, "a", "b", backfillTS2.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
 
 	// Verify that spans signalling backfills at earlier timestamp are allowed.
-	testBoundarySpan(t, ctx, f, "b", "c", backfillTS1.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
+	testBoundarySpan(t, f, "b", "c", backfillTS1.Prev(), jobspb.ResolvedSpan_BACKFILL, statementTime)
 
 	// Verify that no other boundary spans at earlier timestamp are allowed.
-	testIllegalBoundarySpan(t, ctx, f, "a", "f", backfillTS1.Prev(), jobspb.ResolvedSpan_RESTART)
-	testIllegalBoundarySpan(t, ctx, f, "a", "f", backfillTS1.Prev(), jobspb.ResolvedSpan_EXIT)
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS1.Prev(), jobspb.ResolvedSpan_RESTART)
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS1.Prev(), jobspb.ResolvedSpan_EXIT)
+
+	// Verify that attempting to signal a boundary at the latest boundary time with a different
+	// boundary type causes an assertion error.
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS2.Prev(), jobspb.ResolvedSpan_RESTART)
+	testIllegalBoundarySpan(t, f, "a", "f", backfillTS2.Prev(), jobspb.ResolvedSpan_EXIT)
 
 	// Forward spans completing first backfill and signalling and completing second backfill.
-	testBackfillSpan(t, ctx, f, "b", "f", backfillTS1, backfillTS1)
-	testBoundarySpan(t, ctx, f, "b", "f", backfillTS2.Prev(), jobspb.ResolvedSpan_BACKFILL, backfillTS2.Prev())
-	testBackfillSpan(t, ctx, f, "a", "f", backfillTS2, backfillTS2)
+	testBackfillSpan(t, f, "b", "f", backfillTS1, backfillTS1)
+	testBoundarySpan(t, f, "b", "f", backfillTS2.Prev(), jobspb.ResolvedSpan_BACKFILL, backfillTS2.Prev())
+	testBackfillSpan(t, f, "a", "f", backfillTS2, backfillTS2)
 
 	// Forward spans signalling a restart is required.
 	restartTS := makeTS(30)
-	testBoundarySpan(t, ctx, f, "a", "b", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, backfillTS2)
-	testBoundarySpan(t, ctx, f, "b", "f", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, restartTS.Prev())
+	testBoundarySpan(t, f, "a", "b", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, backfillTS2)
+	testBoundarySpan(t, f, "b", "f", restartTS.Prev(), jobspb.ResolvedSpan_RESTART, restartTS.Prev())
 
 	// Simulate restarting by creating a new frontier with the initial highwater
 	// set to the previous frontier timestamp.
@@ -144,13 +149,13 @@ func TestCoordinatorFrontier(t *testing.T) {
 	require.NoError(t, err)
 
 	// Forward spans representing post-restart backfill.
-	testBackfillSpan(t, ctx, f, "a", "b", restartTS, initialHighwater)
-	testBackfillSpan(t, ctx, f, "e", "f", restartTS, initialHighwater)
-	testBackfillSpan(t, ctx, f, "b", "e", restartTS, restartTS)
+	testBackfillSpan(t, f, "a", "b", restartTS, initialHighwater)
+	testBackfillSpan(t, f, "e", "f", restartTS, initialHighwater)
+	testBackfillSpan(t, f, "b", "e", restartTS, restartTS)
 
 	// Forward spans signalling an exit is required.
 	exitTS := makeTS(40)
-	testBoundarySpan(t, ctx, f, "a", "f", exitTS.Prev(), jobspb.ResolvedSpan_EXIT, exitTS.Prev())
+	testBoundarySpan(t, f, "a", "f", exitTS.Prev(), jobspb.ResolvedSpan_EXIT, exitTS.Prev())
 }
 
 type frontier interface {
@@ -162,12 +167,7 @@ type frontier interface {
 }
 
 func testBackfillSpan(
-	t *testing.T,
-	ctx context.Context,
-	f frontier,
-	start, end string,
-	ts hlc.Timestamp,
-	frontierAfterSpan hlc.Timestamp,
+	t *testing.T, f frontier, start, end string, ts hlc.Timestamp, frontierAfterSpan hlc.Timestamp,
 ) {
 	backfillSpan := makeResolvedSpan(start, end, ts, jobspb.ResolvedSpan_NONE)
 	require.True(t, f.InBackfill(backfillSpan))
@@ -178,7 +178,6 @@ func testBackfillSpan(
 
 func testBoundarySpan(
 	t *testing.T,
-	ctx context.Context,
 	f frontier,
 	start, end string,
 	boundaryTS hlc.Timestamp,
@@ -213,7 +212,6 @@ func testBoundarySpan(
 
 func testIllegalBoundarySpan(
 	t *testing.T,
-	ctx context.Context,
 	f frontier,
 	start, end string,
 	boundaryTS hlc.Timestamp,

